python/cudf/bindings/sort.pyx

```diff
@@ -74,3 +72,128 @@ cpdef digitize(column, bins, right=False):
 
     check_gdf_error(result)
     return out
+
+
+def apply_segsort(col_keys, col_vals, segments, descending=False,
+                  plan=None):
+    """Inplace segemented sort
+
+    Parameters
+    ----------
+    col_keys : Column
+    col_vals : Column
+    segments : device array
+    """
+    # prepare
+    nelem = len(col_keys)
+    if nelem == segments.size:
+        # As many seguments as there are elements.
+        # Nothing to do.
+        return
+
+    if plan is None:
+        plan = SegmentedRadixSortPlan(
+            nelem,
+            col_keys.dtype,
+            col_vals.dtype,
+            descending=descending
+        )
+
+    plan.sort(
+        segments,
+        col_keys,
+        col_vals
+    )
+    return plan
+
+
+class SegmentedRadixSortPlan(object):
+    def __init__(self, nelem, key_dtype, val_dtype, descending=False):
+        begin_bit = 0
+        self.sizeof_key = key_dtype.itemsize
+        self.sizeof_val = val_dtype.itemsize
+        end_bit = self.sizeof_key * 8
+        cdef uintptr_t plan
+        cdef size_t c_nelem = nelem
+        cdef int c_descending = descending
+        cdef unsigned c_begin_bit = begin_bit
+        cdef unsigned c_end_bit = end_bit
+        with nogil:
+            plan = <uintptr_t>gdf_segmented_radixsort_plan(
+                c_nelem,
+                c_descending,
+                c_begin_bit,
+                c_end_bit
+            )
+        self.plan = int(plan)
+        self.nelem = nelem
+        self.is_closed = False
+        self.setup()
+
+    def __del__(self):
+        if not self.is_closed:
+            self.close()
+
+    def close(self):
+        cdef uintptr_t c_plan = self.plan
+        with nogil:
+            gdf_segmented_radixsort_plan_free(
+                <gdf_segmented_radixsort_plan_type*>c_plan
+            )
+        self.is_closed = True
+        self.plan = None
+
+    def setup(self):
+        cdef uintptr_t c_plan = self.plan
+        cdef size_t c_sizeof_key = self.sizeof_key
+        cdef size_t c_sizeof_val = self.sizeof_val
+        with nogil:
+            gdf_segmented_radixsort_plan_setup(
+                <gdf_segmented_radixsort_plan_type*>c_plan,
+                c_sizeof_key,
+                c_sizeof_val
+            )
+        self.plan = int(c_plan)
+
+    def sort(self, segments, col_keys, col_vals):
+        cdef gdf_column* c_col_keys = column_view_from_column(col_keys)
+        cdef gdf_column* c_col_vals = column_view_from_column(col_vals)
+    
+        seg_dtype = np.uint32
+        segsize_limit = 2 ** 16 - 1
```

Comment:
Why is the `segsize_limit` 2**16? Probably need to see the C++ side, but would’ve thought its `2**32-1`